src/hotspot/share/prims/universalUpcallHandler.cpp

```diff
@@ -71,6 +72,93 @@ void ProgrammableUpcallHandler::detach_thread(Thread* thread) {
   vm->functions->DetachCurrentThread(vm);
 }
 
+// modelled after JavaCallWrapper::JavaCallWrapper
+Thread* ProgrammableUpcallHandler::on_entry(OptimizedEntryBlob::FrameData* context) {
+  JavaThread* thread = maybe_attach_and_get_thread(&context->should_detach)->as_Java_thread();
+  context->thread = thread;
+
+  assert(thread->can_call_java(), "must be able to call Java");
+
+  // Allocate handle block for Java code. This must be done before we change thread_state to _thread_in_Java,
+  // since it can potentially block.
+  context->new_handles = JNIHandleBlock::allocate_block(thread);
+
+  // After this, we are official in Java Code. This needs to be done before we change any of the thread local
```

Comment:
typo: s/official/officially/
(I see this was copied from Javacalls)

Reply:
Thanks, I wasn't sure whether to leave the comments in or not.